misc/python/materialize/mzexplore/sql.py

```diff
@@ -117,3 +118,41 @@ def try_mzfmt(sql: str) -> str:
         return result.stdout.decode("utf-8").rstrip()
     else:
         return sql.rstrip().rstrip(";")
+
+
+def identifier(sql):
+    """
+    A version of pg8000.native.identifier (1) that is _ACTUALLY_ compatible with
+    the Postgres code (2).
+
+    1. https://github.com/tlocke/pg8000/blob/017959e97751c35a3d58bc8bd5722cee5c10b656/pg8000/converters.py#L739-L761
+    2. https://github.com/postgres/postgres/blob/b0f7dd915bca6243f3daf52a81b8d0682a38ee3b/src/backend/utils/adt/ruleutils.c#L11968-L12050
+    """
+    if not isinstance(sql, str):
+        raise InterfaceError("identifier must be a str")
+
+    if len(sql) == 0:
+        raise InterfaceError("identifier must be > 0 characters in length")
+
+    def is_alpha(c: str) -> bool:
+        return ord(c) >= ord("a") and ord(c) <= ord("z") or c == "_"
+
+    def is_alpanum(c: str) -> bool:
+        return is_alpha(c) or ord(c) >= ord("0") and ord(c) <= ord("9")
+
+    quote = not (is_alpha(sql[0]))
```

Comment:
Note this won't be sufficient in all cases. An identifier composed of all alphanumeric characters may need quoting if it is a reserved keyword (e.g., `FROM`). Materialize is pretty permissive about this, but there are a few cases where we _have_ to treat reserved keywords as keywords and not identifiers if unquoted.

Reply:
Good catch, I guess it was too late for me to double-check this. I added the list of keywords as a soft link under the `sql` folder and an extra check that the identifier string `s` is not a keyword

```python
if s.upper() in keywords():
    quote = True
```
where
```python
@functools.lru_cache(maxsize=1)
def keywords() -> set[str]:
    """
    Return a list of keywords reserved by Materialize.
    """
    with resource_path("sql/keywords.txt").open() as f:
        return set(
            line.strip().upper()
            for line in f.readlines()
            if not line.startswith("#") and len(line.strip()) > 0
        )
```